File: app/src/main/java/me/raatiniemi/worker/presentation/service/ClockOutService.java
/*
 * Copyright (C) 2016 Worker Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.presentation.service;

import android.content.Intent;
import android.util.Log;

import java.util.Date;

import me.raatiniemi.worker.R;
import me.raatiniemi.worker.domain.interactor.ClockOut;
import me.raatiniemi.worker.presentation.notification.ErrorNotification;

public class ClockOutService extends OngoingService {
    private static final String TAG = "ClockOutService";

    public ClockOutService() {
        super(TAG);
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        long projectId = getProjectId(intent);

        try {
            ClockOut clockOut = buildClockOutUseCase();
            clockOut.execute(projectId, new Date());

            dismissPauseNotification(projectId);
            updateUserInterface(projectId);
        } catch (Exception e) {
            Log.w(TAG, "Unable to clock out project: " + e.getMessage());

            sendErrorNotification(projectId);
        }
    }

    protected ClockOut buildClockOutUseCase() {
        return new ClockOut(getTimeRepository());
    }

    private void dismissPauseNotification(long projectId) {
        dismissNotification(projectId);
    }

    private void sendErrorNotification(long projectId) {
        sendNotification(
                projectId,
                ErrorNotification.build(
                        this,
                        getString(R.string.error_notification_clock_out_title),
                        getString(R.string.error_notification_clock_out_message)
                )
        );
    }
}

File: app/src/main/java/me/raatiniemi/worker/features/ongoing/service/ClockOutService.kt
/*
 * Copyright (C) 2019 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.features.ongoing.service

import android.content.Intent
import me.raatiniemi.worker.domain.project.model.Project
import me.raatiniemi.worker.domain.project.usecase.GetProject
import me.raatiniemi.worker.domain.time.Milliseconds
import me.raatiniemi.worker.domain.timeinterval.usecase.ClockOut
import me.raatiniemi.worker.domain.timeinterval.usecase.InactiveProjectException
import me.raatiniemi.worker.monitor.analytics.Event
import me.raatiniemi.worker.monitor.analytics.UsageAnalytics
import org.koin.android.ext.android.inject
import timber.log.Timber

internal class ClockOutService : OngoingService("ClockOutService") {
    private val usageAnalytics: UsageAnalytics by inject()
    private val getProject: GetProject by inject()
    private val clockOut: ClockOut by inject()

    override fun onHandleIntent(intent: Intent?) {
        try {
            val projectId = getProjectId(intent)
            val project = getProject(projectId)

            clockOut(project)

            updateUserInterface(project)
            dismissNotification(project)
        } catch (e: InactiveProjectException) {
            Timber.w(e, "Clock out service called with inactive project")
        } catch (e: Exception) {
            Timber.e(e, "Unable to clock out project")
        }
    }

    private fun clockOut(project: Project) {
        clockOut(project, Milliseconds.now)

        usageAnalytics.log(Event.NotificationClockOut)
    }
}
